Allow toplevel value to be used for preprocessing

Kakadu opened the pull request:
I was building a custom js_of_ocaml toplevel recently and I needed to embed PPX syntax extension. By default it uses system call to invoke preprocessor which doesn't suite js_of_ocaml. The patch allows toplevel values of type `Ast_mapper.mapper` to be used while preprocessing the source.

P.S. Happy New Year :)

gasche commented:
cc @diml @whitequark @alainfrisch : what do you think?

whitequark commented:
I acknowledge the usefulness of such a feature (even without js_of_ocaml!), but I have a few concerns. The main one is that we will have two distinct ways of adding a ppx. How do they interact with each other? This is especially important, given that I think I'd like to see ocamlfind activating extensions like this.

The minor ones are that the messages are inconsistent with the behavior of `#ppx`, the name `#ppx_mapper` does nothing to distinguish itself from `#ppx` in a useful way, `#ppx_mappers_clear` has no equivalent for the out-of-process mappers, and `#d...` directives are very awkwardly named.

I propose to rename `#ppx_mapper` to something that better describes its function (but I am unsure what: `#inproc_ppx`? `#internal_ppx`?); remove `#ppx_mappers_clear` or add a `#ppx` equivalent (I do not have an opinion on which option is preferable); and remove the `#d...` directives, since their full functionality is exposed from the command line and can also be accessed if you `#require "compiler-libs.common"`; it's a niche feature with fairly obtuse output which, I think, doesn't warrant a directive.

Kakadu commented:
>  The main one is that we will have two distinct ways of adding a ppx. How do they interact with each other? This is especially important, given that I think I'd like to see ocamlfind activating extensions like this.

At the moment toplevel applies all external ppx preprocessors and ad-hoc preprocessors after that. Demo output looks like this:

```
➜  asp  ocaml
        OCaml version 4.03.0+dev12-2015-11-20

Findlib has been successfully loaded. Additional directives:
  #require "package";;      to load a package
  #list;;                   to list the available packages
  #camlp4o;;                to load camlp4 (standard syntax)
  #camlp4r;;                to load camlp4 (revised syntax)
  #predicates "p,q,...";;   to set these predicates
  Topfind.reset();;         to force that packages will be reloaded
  #thread;;                 to enable threads

#require "compiler-libs.common";;
/home/kakadu/.opam/4.03.0+trunk+fastppx/lib/ocaml/compiler-libs: added to search path
/home/kakadu/.opam/4.03.0+trunk+fastppx/lib/ocaml/compiler-libs/ocamlcommon.cma: loaded
#require "ppx_const";;
# /home/kakadu/.opam/4.03.0+trunk+fastppx/lib/ppx_const: added to search path
/home/kakadu/.opam/4.03.0+trunk+fastppx/lib/ppx_const/./ppx_const: activated
open Parsetree;;
# open Ast_mapper;;
# 
let dumb_mapper =
  { default_mapper with expr = fun mapper e ->
      match e with
      | {pexp_desc=Pexp_constant (PConst_char 'a');_} as e ->
          {e with pexp_desc=Pexp_constant (PConst_char 'A')}
      | e -> default_mapper.expr mapper e
    };;
val dumb_mapper : Ast_mapper.mapper =
  {attribute = <fun>; attributes = <fun>; case = <fun>; cases = <fun>;
   class_declaration = <fun>; class_description = <fun>; class_expr = <fun>;
   class_field = <fun>; class_signature = <fun>; class_structure = <fun>;
   class_type = <fun>; class_type_declaration = <fun>;
   class_type_field = <fun>; constructor_declaration = <fun>; expr = <fun>;
   extension = <fun>; extension_constructor = <fun>;
   include_declaration = <fun>; include_description = <fun>;
   label_declaration = <fun>; location = <fun>; module_binding = <fun>;
   module_declaration = <fun>; module_expr = <fun>; module_type = <fun>;
   module_type_declaration = <fun>; open_description = <fun>; pat = <fun>;
   payload = <fun>; signature = <fun>; signature_item = <fun>;
   structure = <fun>; structure_item = <fun>; typ = <fun>;
   type_declaration = <fun>; type_extension = <fun>; type_kind = <fun>;
   value_binding = <fun>; value_description = <fun>; with_constraint = <fun>}
# #dsource true;;
# let f () = if%const true then fun () -> 'a' else fun () -> 'b';;

let f () () = 'a';;
val f : unit -> unit -> char = <fun>
# #ppx_mapper dumb_mapper;;

#ppx_mapper  dumb_mapper;;
Extension is added. There are 1 extensions at the moment
# let f () = if%const true then fun () -> 'a' else fun () -> 'b';;
Applying mapper

let f () () = 'A';;
val f : unit -> unit -> char = <fun>
```

> The minor ones are that the messages are inconsistent with the behavior of #ppx, the name #ppx_mapper does nothing to distinguish itself from #ppx in a useful way, #ppx_mappers_clear has no equivalent for the out-of-process mappers, and #d... directives are very awkwardly named.

`#print_parsedtree`, `#print_source` ?

> I propose to rename #ppx_mapper to something that better describes its function (but I am unsure what: #inproc_ppx? #internal_ppx?); remove #ppx_mappers_clear or add a #ppx equivalent (I do not have an opinion on which option is preferable); 

Maybe `fast_ppx` (the same as a file created for this stuff). Adding the directive `#ppx_clear` should not be difficult.

> ... and remove the #d... directives, since their full functionality is exposed from the command line and can also be accessed if you #require "compiler-libs.common"; it's a niche feature with fairly obtuse output which, I think, doesn't warrant a directive.

I like that I am able to enable/disable source output when I want it. Also, `#recflags` is already supported by the toplevel and these  three options are not worse than then `#recflags`.

whitequark commented:
> Maybe fast_ppx

That's completely uninformative.

> these three options are not worse than then #recflags.

```
# #recflags;;
Unknown directive `recflags'.
```

Did you mean `#rectypes`? `#rectypes` is useful for more than debugging; also, it's exposed in a way different than what you propose (it doesn't accept an argument).

I still really dislike that we have two completely independent ways of applying ppx extensions. How about refactoring the codepath behind `#ppx` so that both out-of-process and in-process rewriters are applied in the order that you add them? And there would also be only one `#ppx_clear`.

whitequark commented:
> I like that I am able to enable/disable source output when I want it.

You can still do that if using `Clflags`, though.

Kakadu commented:
Yeah, `#recflags`.  

> How about refactoring the codepath behind #ppx so that both out-of-process and in-process rewriters are applied in the order that you add them?

I will take a look

alainfrisch commented:
I share all @whitequark's comments.

Having a single set of ppxs to be applied (with two different calling modes) seems much better.  Cookies should be properly threaded between all of them (they can be used to pass information from one to the other).

Instead of taking a Ast_mapper.mapper value, one could consider relying on the registration mechanism in Ast_mapper, i.e. simply load a .cmo/.cma file implementing the ppx logic and registering the mapper through Ast_mapper.register, and then use a toplevel directive with string argument(s) to refer to registered mapper (and pass them arguments).  This would make it easier to reuse existing code of ppx that don't necessarily export the mapper value.

Kakadu committed:
Directives to print parsed tree, typed tree and source

Kakadu committed:
Allow to specify value of type Ast_mapper.mapper as PPX extension

Kakadu commented:
Folks, do you think that changing type clflags.mli to 

```
+type ppx_cfg = ExternalPPX of string | LocalPPX of Obj.t
+
...
-and all_ppx = ref ([] : string list)        (* -ppx *)
+and all_ppx = ref ([] : ppx_cfg list)        (* -ppx *)
```

will be bad idea?

alainfrisch commented:
> will be bad idea?

Yes, indeed.  I'd find much better to keep using Clflags.all_ppx to represent command-line parameters and introduce another list (e.g. in Pparse) to represent the two possible kinds of ppx calling mode.  There, instead of Obj.t, you could directly refer to the actual type of mapper (or just a local record of mapping functions).

gasche commented:
Note that it should be possible to give the two kinds of ppx the same type by using a GADT and existential packing. You should be able to reuse the type ['a ast_kind](https://github.com/gasche/ocaml/blob/remove-obj-driver/driver/pparse.mli#L24-L26) that I developped (not merged in trunk) for #351 (note: I'm interested in feedback on this PR):

``` ocaml
type ppx_transform = Ppx_val : 'a ast_kind * ('a -> 'a) -> ppx_transform
```

Having a single sequence of ppx transformers (even if some are skipped dynamically when processing a certain kind of ASTs) would make it easier to ensure that the pass ordering is preserved.

alainfrisch commented:
I'm not sure to see what you mean with the existential.  Registered PPXs all have to support both kinds of AST types (signatures/structures), not only one of them.  (Well, technically, in the context of the toplevel, in-process mappers don't need to deal with signature, but let's be more general to allow linking special compiler drivers with built-in mappers.)   The two kinds of PPXs are those implemented by an external command (which can be abstracted as "string -> string -> unit", the two strings representing the input and output files) and those implemented by an in-process mapper (of type Ast_mapper.mapper).

Kakadu committed:
Moving information about enabled PPX preprocessors to pparse.ml.

From now on Clflags.all_ppx only stores options after parsing command
line arguments.

Some values are exposed in pparse.mli to manage PPX preprocessors:
  val clear_ppx : unit -> unit
  (* disable all ppxs *)
  val add_external_ppx : path:string -> unit
  (* Add PPX specified by command line option *)
  val add_in_process_ppx : Ast_mapper.mapper -> unit
  (* Add ast mapper specified in the toplevel *)

damiendoligez commented:
I think the name should be `plugin_ppx` for this feature.

Kakadu committed:
Added test for new PPX plugins

Kakadu commented:
Okay. 
1. I renamed directive as Damien recommended
2. I left dump options as they are.
3. I adopted gasche's GPR
4. After parsing command line options they are passed to `pparse.ml` and don't get away from there.
5. In process PPX objects are added to `pparse.ml` too.
6. The implementation in `pparse.ml` seems to be very straightforward. I don't understrand why we need new GADT there or why should we abstract everything as function
7. Test added

whitequark commented:
Looks good to me otherwise--though there are a few leftover changes from the rebase (`Fastppx` reference and makefile dep) and a bunch of unrelated commits pulled in before that.

Kakadu commented:
I probably should rebase everything after  release of 4.03...

Kakadu commented:
> Fastppx reference and makefile dep

Are you reading outdated diff, @whitequark  ? Don't see anything like that.

whitequark commented:
@Kakadu I was reading individual commits.

Kakadu commented:
I will be glad to have `approved` tag on this.

Drup commented:
Note that the js_of_ocaml usecase is pretty much resolved (https://github.com/ocsigen/js_of_ocaml/issues/394).

gasche commented on driver/compenv.ml:
```diff
@@ -520,7 +520,9 @@ let readenv ppf position =
-  all_ppx := !last_ppx @ !first_ppx
+  all_ppx := !last_ppx @ !first_ppx;
+  Pparse.clear_ppx();
+  List.iter (fun path ->Pparse.add_external_ppx ~path) !all_ppx
```
space missing: `-> Pparse`

alainfrisch commented on toplevel/toploop.ml:
```diff
@@ -359,6 +359,7 @@ let use_print_results = ref true
 let preprocess_phrase ppf phr =
   let phr =
     match phr with
+    | Ptop_def [] -> phr
```
Why is this special case now required?

gasche commented:
The code seems un-invasive now and I think @Kakadu did good work heeding the comments and advice. I would be willing to merge it provided I have confirmation/approval from either @whitequark or @alainfrisch that know the problem domain better.

whitequark commented on toplevel/topdirs.ml:
```diff
@@ -401,6 +401,52 @@ let _ = add_directive "remove_printer"
+      let mapper = eval_path !toplevel_env path in
+      Pparse.add_in_process_ppx (Obj.obj mapper)
+    else
+      fprintf ppf "type of the value should be 'Ast_mapper.mapper'. Nothing added.\n%!"
```
Nit: `type` should be capitalized.

alainfrisch commented on toplevel/topdirs.ml:
```diff
@@ -401,6 +401,52 @@ let _ = add_directive "remove_printer"
+  try
+    let (path, desc) = Env.lookup_value lid !toplevel_env in
+    let open Path in
+    let rec is_good_type t =
```
This seems a rather fragile way to match on types.  You should do as for printer registration (see e.g. `match_simple_printer_type`), using proper unification.

alainfrisch commented on toplevel/topdirs.ml:
```diff
@@ -401,6 +401,52 @@ let _ = add_directive "remove_printer"
+          syntax tree to specified PPX mappers.";
+    }
+
+let _ = add_directive "ppxs_clear"
```
Is there an actual use case for that?  Does it remove ppx passed on the command-line or through OCAMLPARAM?

whitequark replied:
I'm concerned about this feature because it will interact badly with rewriters registered through `#require` directives; there is essentially no way to bring them back. How about `#remove_in_process_ppx`, which should be entirely sufficient for testing?

Kakadu replied:
Added in fa8e219a5e25b3ecdeca1bc8302edc5a6ec5bc6e

whitequark commented on toplevel/topdirs.ml:
```diff
@@ -401,6 +401,52 @@ let _ = add_directive "remove_printer"
+          syntax tree through the preprocessor command.";
+    }
+
+let _ = add_directive "plugin_ppx"
```
I disagree with @damiendoligez here, `#plugin_ppx` is a rather bad name for this feature because OCaml plugins are the `cmxs` files. Why not just `#in_process_ppx`?

Kakadu replied:
Added in fa8e219a5e25b3ecdeca1bc8302edc5a6ec5bc6e

Kakadu committed:
Reworked type checking for in-process-ppx as @alainfrisch recommended.

Kakadu committed:
Added directive for removing in-process ppxes

whitequark commented on toplevel/topdirs.ml:
```diff
@@ -447,6 +446,13 @@ let _ = add_directive "ppxs_clear"
+    {
+      section = section_options;
+      doc = "Forget all added in process PPX preprocessors";
+    }
```
No no, this is just as bad. What happens once ocamlfind starts to add ppx in-process? I meant that `#remove_in_process_ppx` would take an argument and just that value would be removed.

Kakadu replied:
@whitequark Do you propose to use path as a name for in-process ppx? And `#remove_in_process_ppx` will take this path to decide which ppx to remove?

whitequark replied:
@Kakadu Just compare them by-value. Though, path is also acceptable, and might be more convenient in some cases.

Kakadu committed:
Store registered in-process PPXs with path.

Add directives to remove in-process ppxs by path.

whitequark commented:
@alainfrisch So, what do you think about the current command set? There are advantages and disadvantages to both removing an in-process ppx by value (as I originally thought) and by path (as implemented).

By value:
- exactly the specified mapper will be removed
- but, if it was irreparably shadowed by loading a newer version of the same .cmo, it cannot be removed

By path:
- it is not exactly certain what will be removed
- but, it allows a workflow where a forgetful programmer would first load a newer .cmo and then remove and re-add the in-process ppx

I find it hard to decide between the two. Removing by value appeals because of how unambiguous it is, but removing by path appeals because I have an attention span of a chipmunk.

whitequark commented on driver/pparse.mli:
```diff
@@ -28,6 +28,19 @@ type 'a ast_kind =
 | Structure : Parsetree.structure ast_kind
 | Signature : Parsetree.signature ast_kind
 
+val clear_ppx : unit -> unit
```
This is not used anymore

Kakadu replied:
@whitequark, https://github.com/ocaml/ocaml/pull/386/files#diff-1955e6ad5eb859327058126966ecd606L523

whitequark replied:
Why is that necessary?

Kakadu replied:
Absence of this line slightly changes semantics. Before my changes `Cflags.all_ppx` was used to store all information about external ppxs and code like `all_ppx :=...` was supposed to remove old information and replace it by new one. Now information about ppx extensions is also stored in `Pparse` module and `Cflags.all_ppx` is kind of .... redundant? Can we get rid of it? What about backward compatibility?

Also if I remove `clear_ppx` call something bad can happen if `readenv` will be called twice. (I think)

whitequark commented on driver/pparse.mli:
```diff
@@ -28,6 +28,19 @@ type 'a ast_kind =
+val clear_ppx : unit -> unit
+(** Forget about all installed rewriters: in-process and external ones *)
+
+val clear_in_process_ppxs: unit -> unit
```
As I've explained before, indiscriminately removing all in-process ppxs is an operation that does not seem too useful and it will actively break ocamlfind if it ever starts installing in-process ppxs (which is something I think is quite useful).

Kakadu replied:
Absence of this option forces user to remember all in-process ppxes he added...

whitequark replied:
How is that different from external ppxs?

whitequark commented on driver/pparse.mli:
```diff
@@ -28,6 +28,19 @@ type 'a ast_kind =
+val clear_in_process_ppxs: unit -> unit
+(** Forget about all installed in-process rewriters *)
+
+val disable_in_process_ppx: Longident.t -> unit
```
This should probably be called `remove_in_process_ppx`; "disable" implies that it can be re-enabled but it's just removed entirely.

alainfrisch commented:
> @alainfrisch So, what do you think about the current command set?

What's the motivation for allowing users to unregister a ppx in the toplevel?  There is no way to unload some units, to unregister at_exit hooks, etc.  Unless there is a strong use case, I'd rather leave this part of the proposal out.

alainfrisch commented:
There is a significant overlap in scope between this one and #647.  I'd like to find a common API that serves both needs (in-process ppx for the toplevel, and custom compiler drivers linked with static Parsetree rewriters), since this is really the same thing.

alainfrisch commented on driver/pparse.ml:
```diff
@@ -112,23 +112,49 @@ let rewrite kind ppxs ast =
+    | External path ->
+       ast
+       |> Ast_mapper.add_ppx_context_str ~tool_name
+       |> rewrite Structure [path]
```
This change has the unfortunate effect of creating more file traffic when multiple external ppx are used.  Consecutive `External`s should be grouped.

alainfrisch commented on driver/pparse.ml:
```diff
@@ -112,23 +112,49 @@ let rewrite kind ppxs ast =
+       |> rewrite Structure [path]
+       |> Ast_mapper.drop_ppx_context_str ~restore
+    | InProcess (_,mapper) ->
+        Ast_mapper.(mapper.structure mapper ast)
```
You should arrange so that the tool_name is made available to the mapper through `Ast_mapper.tool_name` (i.e. setting `Ast_mapper.tool_name_ref`).

Kakadu replied:
@alainfrisch, hmm, and which tool is used? For one hand it should be `toplevel`, for another hand it should be stringification of `Longident.t` used for installing in-process ppx.

alainfrisch replied:
I don't understand what you mean.  The same `tool_name` (normally the toplevel, although one could imagine using InProcess mappers in other custom drivers) should simply be installed so that the mapper can have access to it.  What do you mean by "the Longident.t used for installing in-process ppx"?

Kakadu replied:
I got a feeling that I should write something like `Ast_mapper.tool_name_ref := Longident.to_string mapper_name` but It seems that you were talking about simple `Ast_mapper.tool_name_ref := "toplevel"`...

alainfrisch replied:
Rather: `Ast_mapper.tool_name_ref := tool_name` (except that the reference is not exposed for now).

Kakadu replied:
Ah, I think I got it.

Kakadu commented:
Also, folks, should I write in-process ppx with dash from English grammar point of view?

Kakadu committed:
Fixed two issues during rewriting in pparse.ml

1) External PPX don't cause any additional IO.
2) Tool name is set when applying in-process PPXs.

Also slightly rewritten code related to applyign rewriters:
function for applying on structures and signatures merged into
single one.

Kakadu commented:
@alainfrisch I think that I fixed your last two remarks in commit b71cdb2443fec094381ce17a21ff3ce58351063a

Also, I did a few tricks with GADT-related `ast_kind` type introduced by @gasche . I hope that I used it as it was supposed to.

Kakadu commented:
Can we do next round of reviewing? I'm slightly worried about GADT code I introduced in     71fd181

Kakadu commented:
@Drup , @gasche, can you please check that I didn't overcomplicate anything in my last commit, please? (if you have some spare time)

gasche commented:
It seems fine to me -- sorry for not replying earlier. In general, with GADT-based APIs, when the code type-checks then it is also reasonable code. (In the same way, patches over the new GADT-based format implementations are much easier to review because type-checking gives much stronger confidence in their correctness than before.)

Kakadu commented:
Can we do one more iteration of review?

gasche commented:
I'm not making decisions on this PR (I'm not qualified), and would rather delegate to @alainfrisch and @whitequark that seems to have opinions about it. They seem rather "meh" on the proposal for now, and I don't think we would merge before a somewhat more enthusiastic consensus emerges.

whitequark commented:
Not much has changed from my POV:
- I would like to see a `#remove_ppx` command to remove an external PPX, like for in-process PPX.
- I still think that `#clear_in_process_ppxs` will be a footgun and/or effectively unusable since it will remove ubiquitously used ppxs, once we have in-process ppx support in ocamlfind.
- `Pparse.clear_ppx` should be removed in favor of something more rigorous, or at least gain a clear rationale for existence.

alainfrisch commented:
I haven't reviewed the new implementation, but I still need to be convinced about the usefulness of unloading a ppx.  You'd typically set up your REPL environment (loading libraries, registering printers, and now installing in-process ppx) after startup and then use it for the current session.  What would be a typical use case for unloading a ppx?

alainfrisch commented:
Latest comments from @whitequark and me were left unaddressed for more than 1Y.  I suspect the OP lost motivation to push the proposal forward, so I'm close this PR for now.
